modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala

```diff
@@ -44,29 +44,67 @@ package object engine {
     * `Running` status.
     */
   def switch(q: EventQueue)(st: Status): Engine[QState] =
-    modify (QState.status.set(_, st)) *> { if (st == Status.Running) next(q) else get }
+    modify(QState.status.set(_, st)) *> {
+      if (st == Status.Running) prime *> execute(q) *> get
+      else get
+    }
+
+  def prime: Engine[Unit] =
+    gets(QState.prime(_)) >>= {
+      case None => unit
+      case Some(qs) => put(qs)
+    }
 
   /**
     * Adds the `Current` `Execution` to the completed `Queue`, makes the next
     * pending `Execution` the `Current` one, and initiates the actual execution.
     *
     * If there are no more pending `Execution`s, it emits the `Finished` event.
     */
-  // XXX: Handle trying to `next` when there are actions pending in Current.
   def next(q: EventQueue): Engine[QState] =
     (gets(QState.next(_)) >>= {
-       // This should only happen when Queue is empty.
+       // No more Executions left
        case None => send(q)(finished)
-       case Some((actions, s)) => put(s) *> execute(q)(actions)
+         // Execution completed, execute next actions
+       case Some(qs) => put(qs) *> execute(q)
      }) *> get
 
   /**
+    * Checks the `Status` is `Running` and executes all actions in the `Current`
+    * `Execution` in parallel. When all are done it emits the `Executed` event.
+    * It also updates the `QState` as needed.
+    */
+  private def execute(q: EventQueue): Engine[Unit] = {
+
+    // Send the expected event when action is executed
+    def act(t: (Action, Int)): Task[Unit] = {
+      val (action, i) = t
+      action >>= {
+        case Result.OK => q.enqueueOne(completed(i))
+        case Result.Error => q.enqueueOne(failed(i))
+      }
+    }
```

Comment:
Prefer `match` over destructuring assignment because you'll get exhaustiveness checking. Also in general use `flatMap` unless `>>=` is clearly better because it's more efficient.

```scala
... = t match {
  case (action, i) =>
    action.flatMap {
      case ...
    }
}
```